soc/arm/silabs_sim3/sim3u/soc.c

```diff
@@ -0,0 +1,128 @@
+		SI32_CLKCTRL_0, SI32_CLKCTRL_A_APBCLKG1_MISC0 | SI32_CLKCTRL_A_APBCLKG1_MISC1);
+	SI32_CLKCTRL_A_enable_ahb_to_dma_controller(SI32_CLKCTRL_0);
+
+	/* Apply pinctrl gpio settings */
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_0, PORTBANK_0_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_0, PORTBANK_0_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_0, PORTBANK_0_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_0, PORTBANK_0_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_0, PORTBANK_0_PINS_ANALOG);
+
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_1, PORTBANK_1_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_1, PORTBANK_1_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_1, PORTBANK_1_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_1, PORTBANK_1_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_1, PORTBANK_1_PINS_ANALOG);
+
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_2, PORTBANK_2_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_2, PORTBANK_2_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_2, PORTBANK_2_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_2, PORTBANK_2_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_2, PORTBANK_2_PINS_ANALOG);
+
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_3, PORTBANK_3_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_3, PORTBANK_3_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_3, PORTBANK_3_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_3, PORTBANK_3_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_3, PORTBANK_3_PINS_ANALOG);
+
+	/* Configure skips */
+	SI32_PBSTD_A_write_pbskipen(SI32_PBSTD_0, PORTBANK_0_SKIPEN_VALUE);
+	SI32_PBSTD_A_write_pbskipen(SI32_PBSTD_1, PORTBANK_1_SKIPEN_VALUE);
+	SI32_PBSTD_A_write_pbskipen(SI32_PBSTD_2, PORTBANK_2_SKIPEN_VALUE);
+	SI32_PBSTD_A_write_pbskipen(SI32_PBSTD_3, PORTBANK_3_SKIPEN_VALUE);
+
+	/* Configure crossbars */
+	SI32_PBCFG_A_enable_xbar0l_peripherals(SI32_PBCFG_0, CROSSBAR_0_CONFIG & 0xFFFFFFFF);
+	SI32_PBCFG_A_enable_xbar0h_peripherals(SI32_PBCFG_0,
+					       (CROSSBAR_0_CONFIG >> 32) & 0xFFFFFFFF);
+	SI32_PBCFG_A_enable_xbar1_peripherals(SI32_PBCFG_0, CROSSBAR_1_CONFIG & 0xFFFFFFFF);
+
+	/* Enable crossbars */
+	SI32_PBCFG_A_enable_crossbar_0(SI32_PBCFG_0);
+	SI32_PBCFG_A_enable_crossbar_1(SI32_PBCFG_0);
+}
+
+static void vddlow_irq_handler(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* nothing to do here, we just don't want any spurious interrupts */
+}
+
+static void vmon_init(void)
+{
+	/* VMON must be enabled for flash write/erase support */
+
+	NVIC_ClearPendingIRQ(VDDLOW_IRQn);
+
+	IRQ_CONNECT(VDDLOW_IRQn, 0, vddlow_irq_handler, NULL, 0);
+	irq_enable(VDDLOW_IRQn);
+
+	SI32_VMON_A_enable_vdd_supply_monitor(SI32_VMON_0);
+	SI32_VMON_A_enable_vdd_low_interrupt(SI32_VMON_0);
+	SI32_VMON_A_select_vdd_standard_threshold(SI32_VMON_0);
+}
```

Comment:
1. Is flash support required for UART operation? As per the PR description
2. IMO we should convert this to a separate sensor driver

Reply:
1. Your post ends mid-sentence and these lines don't have anything related to flash, but: The flash driver itself is not needed for UART, but: 
- I couldn't find a way to compile without it due to zephyr using flash nodes to figure out how to link the binary
- Changing the AHB clock speed requires changing a FLASHCTRL register: https://github.com/zephyrproject-rtos/zephyr/pull/64209/files/861b3c678668d5157c6ba8384a156c0e036fd678#diff-2e890857b384cfc0e5d42940be13fd38adc48f1acc2d3c260f44d3f1db14ffafR86

Reply:
> 1. don't have anything related to flash

What about `/* VMON must be enabled for flash write/erase support */`?

Reply:
Oh sry, I forgot about that. I put this here because it's required to read/write flash and I assumed that everyone would want this (which may not be true ofc).

I don't think the sensor subsystem is a good fit for VMON because this hardware can only produce notifications in form of interrupts, you can't read the voltage level(it's not an ADC). It also seems to be wired up to other peripherals internally. My assumption is that the flash silicon wants to prevent write write-operation failures by detecting VDD issues.